descartes_moveit/src/seed_search.cpp

```diff
@@ -0,0 +1,455 @@
+    : initial_joint_positions_(initial_joint_pos)
+    , bij_(bij)
+    , min_angle_(min_angle)
+    , step_angle_(step_angle)
+  {}
+
+  seed::JointConfig toJointConfig(const ComboHandle& handle) const
+  {
+    seed::JointConfig result;
+    result.resize(bij_.size());
+
+    std::size_t i;
+    for (i = 0; i < handle.size(); i++)
+    {
+      result[bij_[i]] = min_angle_ + handle[i] * step_angle_;
+    }
+
+    for (; i < bij_.size(); ++i)
+    {
+      result[bij_[i]] = initial_joint_positions_[bij_[i]];
+    }
+
+    return result;
+  }
+
+private:
+  std::vector<double> initial_joint_positions_;
+  BijectionVec bij_;
+  double min_angle_;
+  double step_angle_;
+};
+
+/**
+ * @brief Forward kinematics helper
+ */
+bool doFK(moveit::core::RobotState& state,
+          const moveit::core::JointModelGroup* group,
+          const std::string& tool,
+          const seed::JointConfig& joint_pose,
+          Eigen::Affine3d& result)
+{
+  state.setJointGroupPositions(group, joint_pose);
+  if (!state.knowsFrameTransform(tool))
+  {
+    ROS_WARN("No transform to this tool frame");
+    return false;
+  }
+
+  if (!state.satisfiesBounds())
+  {
+    ROS_WARN("Joint angles do not satisfy robot bounds");
+    return false;
+  }
+
+  result = state.getFrameTransform(tool);
+  return true;
+}
+
+/**
+ * @brief Inverse kinematics helper. Returns the solution in the result parameter.
+ *        May return false if IK failed.
+ */
+bool doIK(moveit::core::RobotState& state,
+          const moveit::core::JointModelGroup* group,
+          const std::string& group_name,
+          const std::string& tool,
+          const Eigen::Affine3d& pose,
+          const seed::JointConfig& seed,
+          seed::JointConfig& result)
+{
+  state.setJointGroupPositions(group_name, seed);
+  if (!state.setFromIK(group, pose, tool, 1, 0.01))
+  {
+    return false;
+  }
+  state.copyJointGroupPositions(group, result);
+  return true;
+}
+
+/**
+ * @brief Returns true if the determinant of the jacobian is not zero.
+ *        Returns false if it is close to zero (and robot is near singularity)
+ */
+bool checkSingularity(moveit::core::RobotState& state,
+                      const moveit::core::JointModelGroup* group)
+{
+ Eigen::Vector3d reference_point_position(0.0, 0.0, 0.0);
+ Eigen::MatrixXd jacobian;
+ state.getJacobian(group, state.getLinkModel(group->getLinkModelNames().back()),
+                   reference_point_position, jacobian);
+
+ return std::abs(jacobian.determinant()) > 0.0001;
+}
+
+// Compares the first n_compare bijection joint values from a and b and tests
+// whether they are within 45 degrees of one another
+inline bool sameJointConfig(const seed::JointConfig& a, const seed::JointConfig& b,
+                            const BijectionVec& bij, unsigned n_compare)
+{
+  for (std::size_t i = 0; i < n_compare; i++)
+  {
+    if (std::fmod(std::abs(a[bij[i]] - b[bij[i]]), 2*M_PI) > M_PI_4) return false;
```

Comment:
I was experimenting with removing equivalent ik solutions (that wrapped beyond a certain range like -PI to PI). I realize that because we don't let the joints wrap it may not actually be an equivalent solution.